### PR TITLE
Only avoid pointer casts when using miri

### PR DESCRIPTION
**Pull request by Darksonn**
Address [this comment](https://github.com/tokio-rs/bytes/pull/542#issuecomment-1113431758). See also https://github.com/rust-lang/rust/pull/96538.

### CODE REVIEW

File: src/bytes.rs [hunk 1]
```diff
@@ -1104,14 +1104,25 @@ unsafe fn release_shared(ptr: *mut Shared) {
...
+}
+
+#[cfg(not(miri))]
+fn ptr_map<F>(ptr: *mut u8, f: F) -> *mut u8
+where
+    F: FnOnce(usize) -> usize,
+{
+    let old_addr = ptr as usize;
+    let new_addr = f(old_addr);
+    new_addr as *mut u8
```
Comment (hawkw) [inline src/bytes.rs:1125]: Perhaps this is a bit pedantic, but this really feels like it ought to have a comment on it, IMO. To someone who doesn't know the specific reason for this, it seems quite suspect that we only do the "correct" thing under Miri, and the "real" code does the thing that's not strict provenance compliant. A comment saying that the thing that's theoretically "correct" under strict provenance actually results in miscompiles in real life, and references the appropriate rustc issue etc.

**Review by hawkw — CHANGES_REQUESTED** [review]
Overall, this seems fine, but I think we really ought to document what's going on here.

File: src/bytes.rs (hunk 1 above)
Comment (Darksonn) [reply src/bytes.rs:1131]: You aren't being pedantic. I added a link to this PR, which has links to the previous PR and the rust issue.

**Review by hawkw — APPROVED** [review]
okay, the comment looks good, thank you!

**Pull request merged**